### PR TITLE
docs: Added MMDS user guide

### LINKED ISSUE #1801: Add MMDS user guide and update also MMDS API docs
MMDS documentation is falling behind, once the new changes targeted on MMDS are on their way to be merged (#1790, #1786, #1772). Dedicated documentation needs to be added, which describes the new changes. MMDS configuration API clarifications needs to be added as well to the firecracker swagger file, in terms of the restrictions related to the IPv4 link-local address range.

### PR DESCRIPTION
**Pull request by iulianbarbu**
## Reason for This PR

Closes #1801.

MMDS documentation is falling behind, once the new changes targeted on MMDS are on their way to be merged (#1790, #1786, #1772). Dedicated documentation was added in this PR, to be reviewed and addressed separated. I chose to add a separate PR with docs because adding it in every PR related to MMDS can not easily scale, and might be conflictual between separate PRs.

## Description of Changes

These changes segregates the information which describes MMDS design from the information which describes its usage.

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The reason for this PR is clearly provided (issue no. or explanation).
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] Any newly added `unsafe` code is properly documented.
- [x] Any API changes are reflected in `firecracker/swagger.yaml`.
- [x] Any user-facing changes are mentioned in `CHANGELOG.md`.

### CODE REVIEW

File: docs/mmds/mmds-user-guide.md [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+```bash
+MMDS_IPV4_ADDR=169.254.170.2
+MMDS_NET_IF=eth0
+ip route add ${MMDS_IPV4_ADDR} dev ${MMDS_NET_IF}
+```
+
+# Inserting and updating metadata
+
+Inserting and updating metadata is possible through the firecracker API server.
+These operations can be issued before and after the microVM is running. To
```
Comment (raduiliescu) [inline docs/mmds/mmds-user-guide.md:73]: I think worth mention this is done from host side, even though is redundant with the fact that is done by Firecracker API server.
Also please replace firecracker with Firecracker.

Please rephrase These operations can be issued.... to An user can create or update the MMDS datastore before the microVM is started or during its operation.

File: docs/mmds/mmds-user-guide.md [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+                  }
+            }      
+    }"
+```
+
+# Retrieving metadata
+
+MicroVM metadata can be retrieved:
+* from host operating system, communicating with the firecracker API server.
+* from guest operating system, communicating directly with MMDS.
```
Comment (raduiliescu) [inline docs/mmds/mmds-user-guide.md:140]: communitating directly with MMDS -> not clear can you also do PUT from microVM, or we accept just GET.
I would rephrase this to something like - from guest OS doing HTTP GET operations to MMDS IP with query ....
Also from host operating system, doing HTTP queries to firecracker API server at  /mmds endpoint.

File: docs/mmds/mmds-user-guide.md [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+               "reservation-id": "r-79054aef"
+          }
+    }
+}  
+```
+
+## Retrieving metadata in the guest operating system
+
+To retrieve existing metadata in guest operating system, a HTTP `GET` request
+must be issued. The requested resource can be referenced by its corresponding
```
Comment (raduiliescu) [inline docs/mmds/mmds-user-guide.md:186]: A HTTP get must be issued to the configured MMDS IP. Worth to add more examples, like getting a leaf, or getting from a list - like Fargate query for container ID where they were getting just a container ID

File: docs/mmds/mmds-user-guide.md [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+
+1. Some agent running on the host sends a `PUT` request with the initial
+   contents of the MMDS, using the Firecracker API. This most likely takes
+   place before the microVM starts running, but may also happen at a later
+   time. Guest MMDS requests which arrive prior to contents being available
+   receive a *NotFound* response.
+1. The contents are saved to the data store.
+1. The guest sends a `GET` request for the secret key, which is intercepted by
+   the device model.
+1. The device model queries the data store, and gets the value associated with
```
Comment (raduiliescu) [inline docs/mmds/mmds-user-guide.md:226]: device model queries the datastore using dumbo stack
A HTTP response is assembled and sent back to the guest, using dumbo stack and device model.

File: src/api_server/swagger/firecracker.yaml [hunk 5]
```diff
@@ -95,17 +95,17 @@ paths:
...
-        description: The id of the guest drive
-        required: true
-        type: string
-      - name: body
-        in: body
-        description: Guest drive properties
-        required: true
-        schema:
-          $ref: "#/definitions/Drive"
+        - name: drive_id
```
Comment (raduiliescu) [inline src/api_server/swagger/firecracker.yaml:98]: I would put the align/whitespace changes from this file in a different commit. But is not a blocking item for this review.

File: SPECIFICATION.md [hunk 6]
```diff
@@ -27,7 +27,9 @@ enough RAM, etc.), customers can rely on the following:
      `<= 5 MiB`. The memory overhead is dependent on the **workload** (e.g. a
```
Comment (raduiliescu) [inline SPECIFICATION.md:27]: This doc change should be mentioned in changelog

File: docs/mmds/mmds-user-guide.md (hunk 3 above)
Comment (iulianbarbu) [reply docs/mmds/mmds-user-guide.md:186]: We can add few other examples to let users know how they can make use of `JSON Pointer` syntax. However, we should make sure users are encouraged to rely on the proper documentation from which they can learn how to reference JSON values using JSON pointers. I think the standard does a better job in this matter, and that is why I would suggest to keep only one simple example in this documentation, to show that the JSON pointer is the actual `GET` request path, and not stress out how to use/form the JSON pointers to reference different values from an example data store. WDYT?

File: docs/mmds/mmds-user-guide.md (hunk 4 above)
Comment (iulianbarbu) [reply docs/mmds/mmds-user-guide.md:226]: I agree, I think this description is too implementation specific. We can also simplify and not mention any of the MMDS internal hooks (device model, dumbo stack etc). From the user perspective, I think it is important what is the flow of this use case (credential rotation), and what input the user needs to provide to make it happen. I would suggest the following rephrasing:

"MMDS receives the `GET` request and in return, it assembles a HTTP response and sends it back to the guest." 

We can simplify further by explaining what we mean by the guest, replacing `guest` with `guest application` or `MMDS client`, to make it more clear where the `GET` request originates from.

What do you think?

File: SPECIFICATION.md (hunk 6 above)
Comment (iulianbarbu) [reply SPECIFICATION.md:27]: It is not clear to me if you are referring to the change below, added by this PR, or to this:
`Firecracker's virtual machine manager threads have a memory overhead
   - `Firecracker's virtual machine manager threads have a memory overhead
     <= 5 MiB. The memory overhead is dependent on the **workload**`

File: docs/mmds/mmds-user-guide.md (hunk 2 above)
Comment (iulianbarbu) [reply docs/mmds/mmds-user-guide.md:140]: This list of ways to retrieve metadata was intended to be a summary of the routes followed by queries, but not actually describing how these queries/requests look like. Related to how this communication looks like, there are two chapters which follow, for HOST and GUEST, with API/requests examples for retrieving metadata from MMDS. If I would rephrase how you suggest, this would mean that I would repeat the same information in the following chapters, because they intend to describe the API and provide examples.

File: docs/mmds/mmds-user-guide.md (hunk 1 above)
Comment (iulianbarbu) [reply docs/mmds/mmds-user-guide.md:73]: Sure, will rephrase as you suggest.

File: SPECIFICATION.md (hunk 6 above)
Comment (raduiliescu) [reply SPECIFICATION.md:27]: We should add in changelog that MMDS documentation was improved and now there are two different sections MMDS Design and MMDS user guide

File: docs/mmds/mmds-user-guide.md (hunk 2 above)
Comment (raduiliescu) [reply docs/mmds/mmds-user-guide.md:140]: I would be specific on what HTTP methods can be used and from where

File: docs/mmds/mmds-user-guide.md (hunk 3 above)
Comment (raduiliescu) [reply docs/mmds/mmds-user-guide.md:186]: I think is important to cover with examples on how MMDS reponds both to get on objects and on leafs

File: docs/mmds/mmds-user-guide.md (hunk 4 above)
Comment (raduiliescu) [reply docs/mmds/mmds-user-guide.md:226]: If we don't specify it here, we should specify it in design so that the dataflow would be clear to all users.

File: src/vmm/src/resources.rs [hunk 7]
```diff
@@ -253,7 +253,7 @@ impl VmResources {
         self.vsock.insert(config)
     }
 
-    /// Settter for mmds config.
+    /// Setter for mmds config.
```
Comment (raduiliescu) [inline src/vmm/src/resources.rs:256]: Add this as a separate commit and/or move it to MMDS code change PR.

File: SPECIFICATION.md (hunk 6 above)
Comment (raduiliescu) [reply SPECIFICATION.md:27]: I don't see the changelog modification.

File: src/api_server/swagger/firecracker.yaml (hunk 5 above)
Comment (lauralt) [reply src/api_server/swagger/firecracker.yaml:98]: +1 on moving these changes to another commit.

File: src/api_server/swagger/firecracker.yaml [hunk 8]
```diff
@@ -328,9 +328,9 @@ paths:
           description: The MMDS configuration as JSON.
           required: true
           schema:
-            type: "#/definitions/MmdsConfig"
+            $ref: "#/definitions/MmdsConfig"
       responses:
-        204:
+        200:
```
Comment (lauralt) [inline src/api_server/swagger/firecracker.yaml:333]: Shouldn't this be `204`? `PUT` requests return `NoContent` status when successful.

File: docs/mmds/mmds-design.md [hunk 9]
```diff
@@ -41,98 +37,38 @@ desired data store structure and contents. Here's a JSON example:
...
-crate used to de/serialize JSON). Most importantly, there is currently no
-maximum size bound for MMDS contents; the users are free to provide arbitrarily
-large inputs. However, the HTTP server is likely to encounter/become a
+The MMDS contents can be updated either via a subsequent `PUT` (that replaces them
+entirely), or using `PATCH` requests, which feed the JSON body into the JSON Merge
+Patch functionality, based on [RFC 7396](https://tools.ietf.org/html/rfc7396). MMDS
+related API requests come from the host, which is considered a trusted environment,
+so there are no checks beside the kind of validation done by HTTP server and
+`serde-json` (the crate used to de/serialize JSON). Most importantly, there is
+currently no maximum size bound for MMDS contents; the users are free to provide
```
Comment (aghecenco) [inline docs/mmds/mmds-design.md:46]: It's also worth mentioning that https://github.com/firecracker-microvm/firecracker/issues/1336 can be hit with large payloads.

File: docs/mmds/mmds-design.md [hunk 10]
```diff
@@ -173,7 +109,8 @@ a reply.
 ### MMDS Network Stack
 
 Somewhat confusingly, this is the name of the component which taps the device
-model. It has hardcoded IP (*169.254.169.254*) and MAC (*06:01:23:45:67:01*)
+model. It has a configured IPv4 address (see [Firecracker MMDS configuration
```
Comment (aghecenco) [inline docs/mmds/mmds-design.md:112]:
```suggestion
model. It has a user-configured IPv4 address (see [Firecracker MMDS configuration
```

File: docs/mmds/mmds-design.md [hunk 11]
```diff
@@ -173,7 +109,8 @@ a reply.
...
+API](../../src/api_server/swagger/firecracker.yaml)) and MAC (*06:01:23:45:67:01*)
```
Comment (aghecenco) [inline docs/mmds/mmds-design.md:113]:
```suggestion
API](../../src/api_server/swagger/firecracker.yaml)) and fixed MAC (`06:01:23:45:67:01`)
```

File: docs/mmds/mmds-user-guide.md [hunk 12]
```diff
@@ -0,0 +1,250 @@
...
+requests. When configuring the microVM, if MMDS needs to be activated, a
+network interface has to be configured to allow MMDS requests. Network
+interface configuration API can be found in the
+[firecracker swagger file](../../src/api_server/swagger/firecracker.yaml).
+
+### Example
+
+```bash
+MMDS_NET_IF=eth0
+curl --unix-socket /tmp/firecracker.socket -i \
```
Comment (aghecenco) [inline docs/mmds/mmds-user-guide.md:20]: nit: align the `\` characters at the same column

File: docs/mmds/mmds-user-guide.md [hunk 13]
```diff
@@ -0,0 +1,250 @@
...
+### Example
+
+```bash
+MMDS_IPV4_ADDR=169.254.170.2
+curl --unix-socket /tmp/firecracker.socket -i \
+    -X PUT "http://localhost/mmds/config" \
+    -H "accept: application/json" \
+    -H "Content-Type: application/json" \
+    -d "{
+             "ipv4_address": "${MMDS_IPV4_ADDR}"
```
Comment (aghecenco) [inline docs/mmds/mmds-user-guide.md:53]: Please keep alignment consistent - I prefer the alignment at line 24 which wastes less space

Comment (aghecenco) [reply docs/mmds/mmds-user-guide.md:53]: This applies to all `curl` invocation examples

File: docs/mmds/mmds-user-guide.md [hunk 14]
```diff
@@ -0,0 +1,250 @@
...
+replace existing metadata, a subsequent HTTP `PUT` request to the `/mmds`
+resource must be issued, using as a payload the new metadata. A complete
+description of metadata insertion firecracker API can be found in the
+[firecracker swagger file](../../src/api_server/swagger/firecracker.yaml).
+
+
+### Example
+
+```bash
+curl --unix-socket /tmp/firecracker.socket -i \
```
Comment (aghecenco) [inline docs/mmds/mmds-user-guide.md:86]: This looks exactly like EC2 IMDS contents, so perhaps it's worth mentioning that the example mimics IMDS behavior. But MMDS isn't restricted to this, users can fill the datastore with whatever is relevant to their workloads.

File: docs/mmds/mmds-user-guide.md [hunk 15]
```diff
@@ -0,0 +1,250 @@
...
+      "guest_mac": "AA:FC:00:00:00:01",
+      "host_dev_name": "tap0"
+      "allow_mmds_requests": true
+    }'
+```
+
+# Configuring the microVM Metadata Service
+
+MMDS can be configured pre-boot only, using the Firecracker API server. This
+can be achieved through a HTTP `PUT` request to `/mmds/config` resource. The
```
Comment (aghecenco) [inline docs/mmds/mmds-user-guide.md:35]:
```suggestion
can be achieved through an HTTP `PUT` request to `/mmds/config` resource. The
```

Comment (aghecenco) [reply docs/mmds/mmds-user-guide.md:35]: This applies to all occurrences of `a HTTP`

File: docs/mmds/mmds-user-guide.md [hunk 16]
```diff
@@ -0,0 +1,250 @@
...
+To retrieve existing MMDS metadata from host operating system, a HTTP `GET`
+request to the `/mmds` resource must be issued. The HTTP response returns the
+existing metadata, as a JSON formatted text. A complete description of
+retrieving metadata Firecracker API can be found in the
+[firecracker swagger file](../../src/api_server/swagger/firecracker.yaml).
+
+### Example
+
+```bash
+curl -s --unix-socket /tmp/firecracker.socket http://localhost/mmds
```
Comment (aghecenco) [inline docs/mmds/mmds-user-guide.md:164]: For consistency, please use a full `curl` invocation (like other examples) - here and in the chapter below too

File: docs/mmds/mmds-design.md (hunk 10 above)
Comment (iulianbarbu) [reply docs/mmds/mmds-design.md]: Done.

File: docs/mmds/mmds-design.md (hunk 11 above)
Comment (iulianbarbu) [reply docs/mmds/mmds-design.md]: Done.

File: docs/mmds/mmds-user-guide.md (hunk 12 above)
Comment (iulianbarbu) [reply docs/mmds/mmds-user-guide.md]: Done.

File: docs/mmds/mmds-user-guide.md (hunk 15 above)
Comment (iulianbarbu) [reply docs/mmds/mmds-user-guide.md]: Done.

File: docs/mmds/mmds-user-guide.md (hunk 14 above)
Comment (iulianbarbu) [reply docs/mmds/mmds-user-guide.md:87]: Done. Mentioned that the metadata must be any valid JSON.

File: docs/mmds/mmds-user-guide.md (hunk 16 above)
Comment (iulianbarbu) [reply docs/mmds/mmds-user-guide.md]: What do you understand by full invocation? Is there anything left in the curl HTTP `GET` requests that you think of?

I've updated the curl examples formatting, adding the `accept` header, to impose to the client what to expect.

File: src/api_server/swagger/firecracker.yaml (hunk 5 above)
Comment (iulianbarbu) [reply src/api_server/swagger/firecracker.yaml:98]: Done.

File: src/api_server/swagger/firecracker.yaml (hunk 8 above)
Comment (iulianbarbu) [reply src/api_server/swagger/firecracker.yaml:333]: You are right, it is actually a `204` NoContent response. Modified to `204`.

File: docs/mmds/mmds-user-guide.md [hunk 17]
```diff
@@ -0,0 +1,256 @@
...
+also the path of the MMDS request. The HTTP response content will contain the
+referenced metadata resource, as a JSON formatted text.
+
+### Example
+
+Retrieving the `latest/meta-data` resource:
+```bash
+MMDS_IPV4_ADDR=169.254.170.2
+RESOURCE_POINTER_OBJ=latest/meta-data
+curl --unix-socket /tmp/firecracker.socket            \
```
Comment (iggy) [inline docs/mmds/mmds-user-guide.md:197]: Using a unix socket in the guest?

File: docs/mmds/mmds-user-guide.md [hunk 18]
```diff
@@ -0,0 +1,256 @@
...
+     http://${MMDS_IPV4_ADDR}/${RESOURCE_POINTER_OBJ} \
+     -H "accept: application/json"
```
Comment (iggy) [inline docs/mmds/mmds-user-guide.md:199]: If this is going to be a required header, I think it should be mentioned in the document.

File: docs/mmds/mmds-user-guide.md [hunk 19]
```diff
@@ -0,0 +1,256 @@
...
+RESOURCE_POINTER_STR=latest/meta-data/ami-id
+curl --unix-socket /tmp/firecracker.socket            \
+     http://${MMDS_IPV4_ADDR}/${RESOURCE_POINTER_STR} \
+     -H "accept: application/json"
+```
+
+Output:
+```json
+"ami-87654321"
+```
```
Comment (iggy) [inline docs/mmds/mmds-user-guide.md:224]: This satisfies the issue I brought up on the slack announcement thread.

File: docs/mmds/mmds-user-guide.md (hunk 17 above)
Comment (iulianbarbu) [reply docs/mmds/mmds-user-guide.md]: Nope, this is a leftover. Removed the `--unix-socket` part.

File: docs/mmds/mmds-user-guide.md (hunk 18 above)
Comment (iulianbarbu) [reply docs/mmds/mmds-user-guide.md:198]: It is not a required header, rather an way to make it clear what content-type would the client need to expect from the MMDS response. 

MMDS sends a HTTP `GET` response with the `Content-Type` header set to `application/json`, without processing the HTTP `GET` request headers sent by the MMDS client. All that it is of interest for the MMDS is the HTTP `GET` request path, which is mapped as a JSON pointer reference. This is already stated in the docs.

File: docs/mmds/mmds-user-guide.md [hunk 20]
```diff
@@ -0,0 +1,299 @@
...
+    "latest": {
+          "meta-data": {
+               "ami-id": "ami-87654321",
+               "reservation-id": "r-79054aef"
+          }
+    }
+}
+```
+
+## Retrieving metadata in the guest operating system
```
Comment (andreeaflorescu) [inline docs/mmds/mmds-user-guide.md:178]: We should call out somewhere that retrieving anything besides string is only supported in the JSON format.

File: docs/mmds/mmds-user-guide.md [hunk 21]
```diff
@@ -0,0 +1,312 @@
...
+
+To retrieve existing MMDS metadata from guest operating system, an HTTP `GET`
+request must be issued. The requested resource can be referenced by its
+corresponding [JSON Pointer](https://tools.ietf.org/html/rfc6901), which is
+also the path of the MMDS request. The HTTP response content will contain the
+referenced metadata resource. The response format can be JSON (experimental) or
+IMDS. The IMDS documentation can be found
```
Comment (raduiliescu) [inline docs/mmds/mmds-user-guide.md:185]: Add something like, one can select the output format by specifying the optional accept header, application/json will format the output to json, while using plain-text or not specifying this optional header will format the output to imds
